src/svc_rqst.c

```diff
@@ -527,33 +525,55 @@ svc_rqst_delete_evchan(uint32_t chan_id)
 int
 svc_rqst_evchan_reg(uint32_t chan_id, SVCXPRT *xprt, uint32_t flags)
 {
+	struct svc_rqst_rec *old_rec = (struct svc_rqst_rec *)xprt->xp_ev;
 	struct svc_rqst_rec *sr_rec;
 	struct rbtree_x_part *t;
-	int code = 0;
+	int code;
 
 	if (chan_id == 0) {
-		__warnx(TIRPC_DEBUG_FLAG_SVC_RQST,
-			"%s: called with chan_id 0, fatal (bug)", __func__);
-		goto out;
+		if (__svc_params->ev_u.evchan.id) {
+			__warnx(TIRPC_DEBUG_FLAG_ERROR,
+				"%s: %p called with chan_id 0, but global/legacy event channel already exists (bug)",
+				__func__, xprt);
+			return (EINVAL);
+		}
+
+		/* Create a global/legacy event channel */
+		code = svc_rqst_new_evchan(&(__svc_params->ev_u.evchan.id),
+					   NULL /* u_data */ ,
+					   SVC_RQST_FLAG_CHAN_AFFINITY);
+		if (code) {
+			__warnx(TIRPC_DEBUG_FLAG_ERROR,
+				"%s: %p failed to create global/legacy channel (%d)",
+				__func__, xprt, code);
+			return (code);
+		}
+
+		chan_id = __svc_params->ev_u.evchan.id;
 	}
 
-	sr_rec = svc_rqst_lookup_chan(chan_id, &t, SVC_XPRT_FLAG_NONE);
+	sr_rec = svc_rqst_lookup_chan(chan_id, &t, SVC_RQST_FLAG_PART_UNLOCK);
 	if (!sr_rec) {
-		mutex_unlock(&t->mtx);
-		code = ENOENT;
-		goto out;
+		__warnx(TIRPC_DEBUG_FLAG_ERROR,
+			"%s: %p unknown chan_id %d",
+			__func__, xprt, chan_id);
+		return (ENOENT);
 	}
 
 	mutex_lock(&xprt->xp_lock);
 
-	if (flags & SVC_RQST_FLAG_XPRT_UREG) {
-		if (chan_id != __svc_params->ev_u.evchan.id) {
-			svc_rqst_evchan_unreg(__svc_params->ev_u.evchan.id,
-					      xprt,
-					      (SVC_RQST_FLAG_LOCKED |
-					       SVC_RQST_FLAG_SREC_LOCKED));
-			svc_xprt_clear(xprt, SVC_XPRT_FLAG_LOCKED);
+	if (old_rec) {
+		if (old_rec->id_k == chan_id) {
+			__warnx(TIRPC_DEBUG_FLAG_SVC_RQST,
+				"%s: %p already registered chan_id %d",
+				__func__, xprt, chan_id);
+			goto unlock;
 		}
+
+		svc_rqst_evchan_unreg(old_rec->id_k, xprt,
+				      (SVC_RQST_FLAG_LOCKED |
+				       SVC_RQST_FLAG_SREC_LOCKED));
```

Comment:
Oh wow. That xprt_unregister() else clause is never executed, because sr_rec->id_k is always set in svc_rqst_new_evchan().

Worse, it could be executed whenever the xprt isn't actually registered yet (xp_ev == NULL); it's code I wrote in c2ea1c6c. Looks like you and I were coding at cross purposes? This needs a re-think. I'll come in early.